plumber/ppl/lib/ppl/retention/record_deleter_queries.ex

```diff
@@ -0,0 +1,50 @@
+defmodule Ppl.Retention.RecordDeleterQueries do
+  @moduledoc """
+  Database queries for the retention record deleter.
+  """
+
+  import Ecto.Query
+
+  alias Ppl.EctoRepo
+  alias Ppl.PplRequests.Model.PplRequests
+
+  @doc """
+  Deletes up to `limit` expired records (where expires_at < now).
+  First deletes associated blocks, then deletes the ppl_requests.
+  Returns the count of deleted records.
+  """
+  @spec delete_expired_batch(non_neg_integer()) :: {:ok, non_neg_integer()} | {:error, term()}
+  def delete_expired_batch(limit) do
+    now = NaiveDateTime.utc_now()
+
+    expired_records =
+      from(pr in PplRequests,
+        where: pr.expires_at < ^now,
+        select: %{id: pr.id, ppl_artefact_id: pr.ppl_artefact_id},
+        limit: ^limit
+      )
+      |> EctoRepo.all()
+
+    case expired_records do
+      [] ->
+        {:ok, 0}
+
+      records ->
+        Enum.each(records, fn %{ppl_artefact_id: ppl_id} ->
+          Block.delete_blocks_from_ppl(ppl_id)
```

Comment:
We should not use here `ppl_artefact_id` since in the case of a partially rebuilt pipeline, it will have the value of the original pipeline's ppl_id, so the blocks from the rebuilt pipeline will be left in the database.

Instead, you should use the `id` field from the select query.